File: benchmark/clnlbeam.jl
# Copyright (c) 2022: Miles Lubin and contributors
#
# Use of this source code is governed by an MIT-style license that can be found
# in the LICENSE.md file or at https://opensource.org/licenses/MIT.

using JuMP
import Ipopt

function solve_clnlbeam(model, N)
    set_optimizer_attribute(model, "max_iter", 3)
    h = 1 / N
    alpha = 350
    @variables(model, begin
        -1 <= t[1:(N+1)] <= 1
        -0.05 <= x[1:(N+1)] <= 0.05
        u[1:(N+1)]
    end)
    @NLobjective(
        model,
        Min,
        sum(
            0.5 * h * (u[i+1]^2 + u[i]^2) +
            0.5 * alpha * h * (cos(t[i+1]) + cos(t[i])) for i in 1:N
        ),
    )
    @NLconstraint(
        model,
        [i = 1:N],
        x[i+1] - x[i] - 0.5 * h * (sin(t[i+1]) + sin(t[i])) == 0,
    )
    @constraint(
        model,
        [i = 1:N],
        t[i+1] - t[i] - 0.5 * h * u[i+1] - 0.5 * h * u[i] == 0,
    )
    optimize!(model)
end

function get_model(arg)
    if arg == "direct"
        return direct_model(Ipopt.Optimizer())
    elseif arg == "no-bridges"
        return Model(Ipopt.Optimizer; add_bridges = false)
    else
        return Model(Ipopt.Optimizer)
    end
end

function main(io::IO, Ns = [5_000, 50_000, 500_000])
    for type in ["direct", "no-bridges", "bridges"]
        for n in Ns
            start = time()
            solve_clnlbeam(get_model(type), n)
            run_time = round(Int, time() - start)
            println(io, "$type clnlbeam-$n $run_time")
        end
    end
end

if isempty(ARGS)
    main(stdout, [5])
else
    open(joinpath(@__DIR__, "benchmarks.csv"), "a") do io
        main(io)
    end
end

File: benchmark/clnlbeam.py
# Copyright (c) 2022: Miles Lubin and contributors
#
# Use of this source code is governed by an MIT-style license that can be found
# in the LICENSE.md file or at https://opensource.org/licenses/MIT.

import pyomo.environ as pyo
from pyomo.opt import SolverFactory
from pyomo.core import cos, sin
import os
import time

def solve_clnlbeam(N):
    model = pyo.ConcreteModel()
    model.N = N
    model.h = 1.0/model.N
    model.VarIdx = pyo.RangeSet(model.N+1)
    model.t = pyo.Var(model.VarIdx, bounds=(-1.0,1.0), initialize=lambda m,i: 0.05*cos(i*m.h))
    model.x = pyo.Var(model.VarIdx, bounds=(-0.05,0.05), initialize=lambda m,i: 0.05*cos(i*m.h))
    model.u = pyo.Var(model.VarIdx, initialize=0.01)
    alpha = 350
    model.c = pyo.Objective(
        expr=sum(0.5*model.h*(model.u[i+1]**2+model.u[i]**2) + 0.5*alpha*model.h*(cos(model.t[i+1])+cos(model.t[i])) for i in model.VarIdx if i != model.N+1)
    )
    def cons1_rule(m, i):
        if i == m.N+1:
            return pyo.Constraint.Skip
        return m.x[i+1] - m.x[i] - (0.5*m.h)*(sin(m.t[i+1])+sin(m.t[i])) == 0
    model.cons1 = pyo.Constraint(model.VarIdx, rule=cons1_rule)
    def cons2_rule(m,i):
        if i == m.N+1:
            return pyo.Constraint.Skip
        return m.t[i+1] - m.t[i] - (0.5*m.h)*m.u[i+1] - (0.5*m.h)*m.u[i] == 0
    model.cons2 = pyo.Constraint(model.VarIdx, rule=cons2_rule)
    opt = SolverFactory('ipopt')
    opt.options["max_iter"] = 3
    opt.solve(model, tee=True)
    return model

def main(Ns = [5_000, 50_000, 500_000]):
    dir = os.path.realpath(os.path.dirname(__file__))
    for n in Ns:
        start = time.time()
        model = solve_clnlbeam(n)
        run_time = round(time.time() - start)
        with open(dir + "/benchmarks.csv", "a") as io:
            io.write("pyomo clnlbeam-%i -1 %i\n" % (n, run_time))
    return

main()
